File: src/main/java/dao/DaoPersona.java
package dao;

import entity.Persona;
import org.jdbi.v3.sqlobject.customizer.BindBean;
import org.jdbi.v3.sqlobject.statement.GetGeneratedKeys;
import org.jdbi.v3.sqlobject.statement.SqlQuery;
import org.jdbi.v3.sqlobject.statement.SqlUpdate;

public interface DaoPersona {

    @SqlUpdate("CREATE TABLE PERSONA(\n" +
            "    a_email varchar(30),\n" +
            "    d_nombre varchar(30),\n" +
            "    a_pass_hasheado varchar(150)\n" +
            ");\n" +
            "\n" +
            "alter table PERSONA ADD CONSTRAINT PK_PERSONA primary key(a_email);\n" +
            "\n" +
            "INSERT INTO PERSONA(a_email, d_nombre, a_pass_hasheado) VALUES('juan@juan.com', 'Juan', '1234');\n" +
            "INSERT INTO PERSONA(a_email, d_nombre, a_pass_hasheado) VALUES('david@david.com', 'David', 'asdf');\n" +
            "\n" +
            "CREATE TABLE PROBLEMATICA(\n" +
            "    c_id SERIAL,\n" +
            "    a_nombre varchar(60),\n" +
            "    a_descripcion varchar(500),\n" +
            "    f_fecha_creacion date\n" +
            ");\n" +
            "\n" +
            "alter table PROBLEMATICA add constraint PK_PROBLEMATICA primary key(c_id);\n" +
            "\n" +
            "CREATE TABLE PERSONA_PROBLEMATICA(\n" +
            "    a_id varchar(40),\n" +
            "    a_email varchar(30),\n" +
            "    c_id_problematica int,\n" +
            "    b_interventor boolean\n" +
            ");\n" +
            "\n" +
            "ALTER TABLE PERSONA_PROBLEMATICA add constraint PK_PERSONA_PROBLEMATICA primary key(a_id);\n" +
            "ALTER TABLE PERSONA_PROBLEMATICA add constraint FK_PERS_PROB_PERS foreign key(a_email) REFERENCES PERSONA(a_email);\n" +
            "ALTER TABLE PERSONA_PROBLEMATICA add constraint FK_PERS_PROB_PROB foreign key(c_id_problematica) REFERENCES PROBLEMATICA(c_id);")
    void crearTable();

    @SqlUpdate("INSERT INTO PERSONA(a_email, d_nombre, a_pass_hasheado) VALUES(:email, :nombre, :pass)")
    @GetGeneratedKeys
    Persona agregarPersona(@BindBean Persona persona);

    @SqlQuery("Select d_nombre, a_email FROM PERSONA WHERE a_email = :email AND a_pass_hasheado = :pass")
    Persona darPersonaPorCredenciales(@BindBean Persona persona);

    @SqlUpdate("DROP SCHEMA public CASCADE;\n" +
            "    CREATE SCHEMA public;\n" +
            "    \n" +
            "    GRANT ALL ON SCHEMA public TO postgres;\n" +
            "    GRANT ALL ON SCHEMA public TO public;")
    void eliminarTablas();
}

File: src/main/java/dao/DaoPersona.kt
package dao

import entity.Persona
import org.jdbi.v3.core.Jdbi
import java.util.*

/**
 * Manejo de excepciones añadido.
 */
class DaoPersona(val jdbi: Jdbi){

    fun agregarPersona(persona: Persona): Persona{
        return jdbi.withHandle<Persona, Exception>{
            it.createUpdate("INSERT INTO PERSONA(a_email, d_nombres, d_apellidos, a_pass_hasheado) VALUES(:email, :nombres, :apellidos, :pass)")
                .bindBean(persona)
                .executeAndReturnGeneratedKeys()
                .mapToBean(Persona::class.java)
                .findOnly()
        }
    }

    fun darPersonaPorCredenciales(persona: Persona): Optional<Persona> {
        return jdbi.withHandle<Optional<Persona>, Exception>{
            it.createQuery("SELECT d_nombres, d_apellidos, a_email FROM PERSONA WHERE a_email = :email AND a_pass_hasheado = :pass")
                .bindBean(persona)
                .mapToBean(Persona::class.java)
                .findFirst()
        }
    }

    fun darPersona(email: String): Optional<Persona>{
        return jdbi.withHandle<Optional<Persona>, Exception> {
            it.createQuery("SELECT * FROM PERSONA WHERE a_email = :email")
                .bind("email", email)
                .mapToBean(Persona::class.java)
                .findFirst()
        }
    }

    fun verificarExistencia(email: String): Boolean {
        return jdbi.withHandle<Boolean, Exception> {
            it.createQuery("SELECT * FROM PERSONA WHERE upper(a_email) = upper(:email)")
                .bind("email", email)
                .mapToBean(Persona::class.java)
                .findFirst()
                .isPresent()
        }
    }

    /**
     * Busca personas por email que no han sido invitadas a la problematica
     * Consulta SQL por Saul Aponte. Profesor de la universidad de Ibagué.
     */
    fun darPersonasNoInvitadas(email: String, emailRemitente: String, idProblematica: Int): List<Persona> {
            return jdbi.withHandle<List<Persona>, java.lang.Exception>{
                it.createQuery("SELECT DISTINCT a_email, d_nombres, d_apellidos " +
                    "FROM persona WHERE a_email like :email and a_email != :emailRemitente " +
                    "and a_email not in (select a_email_remitente from invitacion where invitacion.c_id_problematica = :idProblematica " +
                    "union all select a_email_destinatario from invitacion where invitacion.c_id_problematica = :idProblematica)")
                    .bind("email", "%$email%" )
                    .bind("idProblematica", idProblematica)
                    .bind("emailRemitente", emailRemitente)
                    .setMaxRows(5)
                    .mapToBean(Persona::class.java)
                    .list()
        }
    }

    fun darInterventores(idProblematica: Int): List<String> {
        return jdbi.withHandle<List<String>, Exception> {
            it.createQuery("""SELECT d_nombre FROM PERSONA_PROBLEMATICA PP  
            WHERE b_interventor = true AND c_id_problematica = :idProblematica
            """.trimIndent())
            .bind("idProblematica", idProblematica)
            .mapTo(String::class.java)
            .list()
        }

    }
}